cdap-explore/src/main/java/co/cask/cdap/explore/executor/QueryExecutorHttpHandler.java

```diff
@@ -201,33 +198,14 @@ public void downloadQueryResults(HttpRequest request, HttpResponder responder,
         return;
       }
 
-      StringBuffer sb = new StringBuffer();
-      sb.append(getCSVHeaders(exploreService.getResultSchema(handle)));
-      sb.append('\n');
-
-      List<QueryResult> results;
-      results = exploreService.previewResults(handle);
-      if (results.isEmpty()) {
-        results = exploreService.nextResults(handle, DOWNLOAD_FETCH_CHUNK_SIZE);
-      }
-
-      ChunkResponder chunkResponder = responder.sendChunkStart(HttpResponseStatus.OK, null);
+      QueryResultsBodyProducer queryResultsBodyProducer = new QueryResultsBodyProducer(exploreService, handle);
+      queryResultsBodyProducer.initialize();
```

Comment:
Why need to separate into two calls? Can't the initialization done in the constructor?